### DISCUSSION TITLE
MQTT Issues - PictureFrame2020

### DISCUSSION THREAD

**Discussion by jimmyg83** [pi3d/pi3d_demos#43] [Q&A]
Hello,

I am having some issues with the MQTT functionality of the PictureFrame2020.py script. I am running this on 2 raspberry pi4's one 4gb and one 8gb and both have the same issue.
I have set it up as per all the instructions etc however I can only get it to work if I run the script in a debug mode (I have been using Thonny and it only works if I use the "nicer" debug function not the "faster" debug function).
I have been playing in the code to try and pin point the issue without luck. It almost looks like the callbacks stop functioning when the code gets past the on_connect callback function. I added a publish script to a status topic on the MQTT and a print in the on_connect and on_message functions to see what was happening. I can see the print and MQTT message from the initial on_connect and then I never get anything from the on_message function (even when the script publishes on the paused topic at the end of the MQTT setup code).

I have also written a standalone MQTT script to test the MQTT functions and I can get this to work without any issues. The rest of the PictureFrame2020 script appears to be working fine just MQTT does not work for me.

Has anyone else had this issue or can point me in the direction of what might be causing it.

Thanks,

**Comment by paddywwoof** [comment]
Hi, if you have folllowed Wolfgangs instructions then presumably you have mosquitto running (and the tests he suggests ran ok). Might as well try to eliminate the variable of running from inside Thonny... What happens if you a) run PictureFrame in a command prompt in a terminal window opened from the desktop?
```bash
$ cd ~/pi3d_demos
$ python3 PictureFrame2020.py --keyboard=true --verbose=true
```
b)  Exit to the command line (i.e. don't have the X11 desktop running at all) then at the command line
```bash
$ cd ~/pi3d_demos
$ sudo xinit /usr/bin/python3 /home/pi/pi3d_demos/PictureFrame2020.py --keyboard=true --verbose=true
```
setting keyboard true will allow you to stop the program with ESC and verbose true will make any error caught in the MQTT section get printed - worth a try.

**Reply by jimmyg83** [reply]
Thank you for your prompt reply!

I have tried running it from command line from both within desktop and from command line boot option also without luck.
I can only get MQTT to work in the debug mode of thonny.
I have even used the autostart service as I had originally setup on my 4gb pi4 with no luck.

I am not using mosquitto but qiot running on my qnap nas which is working fine controlling/monitoring a magnitude of other sensors/devices. I had written a straight MQTT script and had this working to turn the screen on/off however I had issues getting the pictureframe2020 to start displaying again when the screen came back on so I started playing with the inbuilt MQTT feature.

I am using verbose and added some additional prints to see when things are happening. It comes up with no errors, it does connect to my MQTT broker and will publish a "connect" message (in on_connect function) however the on_message callback function never gets called. I can add publish code into the main display loop and this works fine but no matter what I do the on_message functions does not get called. Even if I publish on a topic that is subscribed too in the code the on_message function does not get called. Well unless I run in debug mode (not sure why this makes a difference).

Not sure where to go from here.

Thanks,

**Reply by paddywwoof** [reply]
Hmm that does sound odd. Is there any way to tell if the subscription is being registered by the broker. i.e. some kind of logging. I think mosquitto has some kind of logging (but not sure if it logs subscriptions) The other thing I might try is increasing QOS to 1 or 2 for a topic to see what that does. Also add an on_subscribe() function and get that to print a confirmation that the topic has been registered by the broker.

Let me know how you get on - sorry I can't be much help. No idea why running in Thonny debug mode would make any difference!

Paddy

**Reply by jimmyg83** [reply]
Thanks Paddy,

I had a bit more of a play and read the documentation for Paho-Mqtt.
I have now resolved my issue.
What fixed it for me was to move the subscribe calls from the main "if config.USE_MQTT" function to inside the "on_connect" callback function. I had read somewhere that if the broker connection gets disconnected that the subscribes would not be re-established unless they are in the on_connect function. I did add an on_disconnect function to see if it was disconnecting however it was never called in my testing but at least it is now working.

Thanks,

**Comment by paddywwoof** [comment]
@jimmyg83 That's great, thanks for the feedback. I will change the PictureFrame2020 code on master next release - without digging into how all the MQTT works it looks like doing the callback registration as part of the on_connect function is the logical place to do it.

Paddy

**Comment by CyclistHomeCoder** [comment]
Would it make sense to separate all MQTT stuff out in a separate file and only import it if needed (configured)?

**Comment by paddywwoof** [comment]
Yes, at some stage the whole thing will need to be refactored with class instances rather than a global scope and functionality split out for the loop control, file loading, image info etc etc. Sometime early next year I will have a closer look. Have a look at this for the kind of way to break it down https://github.com/helgeerbe/picture_frame, https://pypi.org/project/picframe/
